### DISCUSSION TITLE
How swc handle UTF-8 string

### DISCUSSION THREAD

**Discussion by Steven19Lee** [swc-project/swc#7701] [Q&A] [answered]
Hi, I am new to Rust and I am reading SWC source code to learn how to build a Rust parser. I encounter some issues with UTF-8 string. It is ok to iterate UTF-8 string with  `chars` iterator, but how can I get a slice of UTF-8 string with some function like `slice(start_index, end_index)` I take look into `swc_common` and using StringInput, I found that it still can not index UTF-8 string correctly, how should i use it for slice a UTF-8 string ? 
```rs
use swc_common::input::{StringInput, Input};
use swc_common::BytePos;
fn main() {
    let utf8_str = "中文字串";
    let mut input =  StringInput::new("中文字串", BytePos(0), BytePos(utf8_str.len().try_into().unwrap()));
    println!("{:?}", input.slice(BytePos(0), BytePos(3))); // want to slice string like input.slice(BytePos(0), BytePos(1)) to get frist utf-8 char not bytes
    println!("{:?}", &utf8_str[0..3]);
}

```

**Comment by kdy1** [comment, marked as answer]
https://github.com/swc-project/swc/blob/11efd4e7c5e8081f8af141099d3459c3534c1e1d/crates/testing/src/lib.rs#L110-L112

https://github.com/swc-project/swc/blob/11efd4e7c5e8081f8af141099d3459c3534c1e1d/crates/swc_ecma_parser/examples/parse-all.rs#L25

**Reply by Steven19Lee** [reply]
Hi @kdy1  thanks for your reply, so SWC already handle UTF8 bytes when loading file ? Because slice method of StringInput seems to slice a &str by BytePos how StringInput should use to handle UTF8 index ?

**Reply by kdy1** [reply]
You should use `chars`. You need to learn more about string representation of rust

**Reply by Steven19Lee** [reply]
@kdy1 Ya I know that I need to use chars iter for get UTF8 char from string , I just want to ask how SWC get correct UTF-8 string by index ~ thanks
